### PR TITLE
add gpu ram collector for nvidia feature flag

### PR DESCRIPTION
**Pull request by jamartin9**
## Description
Adds multiple GPU VRAM support with styling by default for up to 7 cards.

Does **NOT** include GPU processes, power usage, core frequency or graph toggles

Extra changes:

1. Show `SWP` only when enabled in basic mode

2. Use `zfs_keys_read` in `get_arc_data` 

3. Toggle `ARC` with label instead of point

Screens:

Windows with swap, single gpu, no arc in basic mode

bottom-gpu

Linux swap, 8 gpus, with arc

linux-bottom

Linux swap, 8 gpus, with arc in basic mode (percentage toggled)

linux-basic

Linux no swap, no gpus, no arc

fedora

Freebsd no swap, no gpu, no arc in basic mode

freebsd-basic

## Issue

1. Mentioned in issue #298 and other issues that GPU stats should include graph toggles, gpu core frequency, gpu processes, gpu power usage, and be hidden by default (like battery).

Which of these are required for this pull request to be merged?

Should the current temperature readings for gpus be hidden? (currently only shown when data is detected at runtime)

2. Mentioned in issue 787 that heim might be removed during refactor, so the duplication of `get_gpu_data` may not matter? (nvml only supports nvidia on windows/linux)

3. Memory bars in basic mode share a layout with network.

Is balancing the widgets drawn between memory and network in basic mode needed? This may complicate the layout if more network features are added in the future. (more than a few gpus is probably uncommon)

## Testing

_If relevant, please state how this was tested. All changes **must** be tested to work:_

_If this is a code change, please also indicate which platforms were tested:_

- [X] _Windows_
- [x] _macOS_
- [X] _Linux_
- [X] _freeBSD_

## Checklist

_If relevant, ensure the following have been met:_

- [X] _Areas your change affects have been linted using rustfmt (`cargo fmt`)_
- [X] _The change has been tested and doesn't appear to cause any unintended breakage_
- [ ] _Documentation has been added/updated if needed (`README.md`, help menu, etc.)_
- [ ] _The pull request passes the provided CI pipeline_
- [X] _There are no merge conflicts_
- [ ] _If relevant, new tests were added (don't worry too much about coverage)_

I did **NOT** test on MAC

I did **NOT** test with a real multi device or **SLI** setup (see https://github.com/jamartin9/bottom/commit/052eee28efb56676e7273ef750ff444d444377cf for the test code of the images that duplicates an existing gpu report with data changed to add 7 gpus for style cycling with defaults)

If any of the Issues should be addressed or a different approach used please let me know.

Feel free to close this PR if it is unwanted.

### CODE REVIEW

**Comment by jamartin9** [comment]
Apologizes for pushing extra commits but I noticed a few bugs. 

Take your time with review!

**Comment by ClementTsang** [comment]
Just a heads up, I'm definitely interested in merging this - just might take me a bit of time to get to reviewing it due to some IRL stuff.

**Comment by ClementTsang** [comment]
Looking through it now, sorry for the delay, and thanks for the updates after some other PRs were merged.

Tested on my M1 macOS and it seems to work fine (that is, it does nothing and bottom continues on working, which seems to make sense in this context).

**Comment by jamartin9** [comment]
Thanks for taking some time to review. 
The latest changes were to resolve the merge conflicts and tidy up the code a bit. I did a quick check of the changes on the linux system with a gpu again. I will try to address comments when time permits.

**Review by ClementTsang — COMMENTED** [review]
Left some comments. For the most part it looks fine, just some suggestions.

File: src/app/data_harvester/memory/general/heim.rs [hunk 1]
```diff
@@ -8,13 +8,19 @@ pub async fn get_mem_data(
...
-        (Ok(None), Ok(None), Ok(None))
+        (Ok(None), Ok(None), Ok(None), Ok(None))
     } else {
-        join!(get_ram_data(), get_swap_data(), get_arc_data())
+        join!(
+            get_ram_data(),
+            get_swap_data(),
+            get_arc_data(),
+            get_gpu_data()
+        )
```
[the remark this answers is not in the document]
Comment (jamartin9) [reply src/app/data_harvester/memory/general/heim.rs]: Changed the tuple to a struct with cfg gated fields

File: src/canvas/widgets/mem_basic.rs [hunk 2]
```diff
@@ -42,12 +26,7 @@ impl Painter {
         }
 
         let ram_ratio = if let Some(mem) = mem_data.last() {
-            mem.1 / 100.0
-        } else {
-            0.0
-        };
-        let swap_ratio = if let Some(swap) = swap_data.last() {
-            swap.1 / 100.0
+            mem.1
```
[the remark this answers is not in the document]
Comment (jamartin9) [reply src/canvas/widgets/mem_basic.rs]: renamed ratio to percentage

File: src/canvas/widgets/mem_basic.rs [hunk 3]
```diff
@@ -56,64 +35,133 @@ impl Painter {
...
+                            .label_style(style)
+                            .gauge_style(style),
+                    );
+                });
+            }
         }
 
+        let constraint_layout: Vec<Constraint> = std::iter::repeat(Constraint::Length(1))
+            .take(draw_widgets.len())
+            .collect();
```
[the remark this answers is not in the document]
Comment (jamartin9) [reply src/canvas/widgets/mem_basic.rs]: Changed to use vec macro. It was a merge oversight from trying to reuse the slice layout with a non const size.

**Comment by jamartin9** [comment]
Thanks for the feedback.
I tried to address your comments. Please let me know if more needs to be done.
For posterity's sake here are some more changes in this pr:
1. build on freebsd again
2. re-add mem basic mode percentage toggle

I gave it another quick build and run on linux w/gpus and freebsd w/o

**Review by ClementTsang — APPROVED** [review]
Generally looks good to me, thanks for the changes!

One more thing I would probably want after some thought is a config option to enable/disable this feature on runtime. If you want, I can add this myself after merging this PR - lmk if you would prefer that.

**Comment by jamartin9** [comment]
> One more thing I would probably want after some thought is a config option to enable/disable this feature on runtime. If you want, I can add this myself after merging this PR - lmk if you would prefer that.

Thanks. I think it would be preferable for ergonomics if you were to add the toggle in another pr. I can assist in testing if needed.

**Comment by ClementTsang** [comment]
Sounds good.

@all-contributors please add @jamartin9 for code.

**Pull request merged**